### PR TITLE
[core][experimental] Raise an exception if a leaf node is found during compilation

### PR DESCRIPTION
**Pull request by kevin85421**
## Why are these changes needed?

Leaf nodes are nodes that are not output nodes and have no downstream nodes. If a leaf node raises an exception, it will not be propagated to the driver. Therefore, this PR raises an exception if a leaf node is found during compilation.

### Another solution: implicitly add leaf node to `MultiOutputNode`

Currently, the function execute can return multiple `CompiledDAGRef`s. The UX we want to provide is to implicitly add leaf nodes to the `MultiOutputNode` but not return the references of the leaf nodes. For example, a `MultiOutputNode` is containing 3 DAG nodes (2 normal DAG nodes + 1 leaf node).

```python
x, y = compiled_dag.execute(input_vals) # We don't return the ref for the leaf node.
```

However, the ref for leaf node will be GC(ed) in execute, and `CompiledDAGRef`’s _del_ will call get if it was never called which makes execute to be a sync instead of an async operation which is not acceptable.

## Checks

- [ ] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
    - [ ] I've added any new APIs to the API Reference. For example, if I added a 
           method in Tune, I've added it in `doc/source/tune/api/` under the 
           corresponding `.rst` file.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

**Comment by rkooo567** [comment]
> Another solution: implicitly add leaf node to MultiOutputNode

Can you create an issue for this? Also can you share me the error message?

File: python/ray/dag/compiled_dag_node.py [hunk 1]
```diff
@@ -946,6 +946,26 @@ def _preprocess(self) -> None:
...
+            ):
+                leaf_nodes.append(task.dag_node)
+        # Leaf nodes are not allowed because the exception thrown by the leaf
+        # node will not be propagated to the driver.
+        if len(leaf_nodes) != 0:
+            raise ValueError(
+                "Compiled DAG doesn't support leaf nodes that don't have "
+                "downstream nodes and are not output nodes. There are "
+                f"{len(leaf_nodes)} leaf nodes in the DAG. Please add them to "
+                f"the MultiOutputNode. These nodes are: {leaf_nodes}"
```
Comment (rkooo567) [inline python/ray/dag/compiled_dag_node.py:966]: Can you improve the error message to show how to solve this error step-by-step? For example, assuming a leaf node is w.f.bind() it could say sth like `add the output of w.f.bind() to MultiOutputNode`

What I recommend you is to try raising the error on your own and fix it looking at this error message. I think it is not very trivial if you assume you are not a developer

**Review by rkooo567 — COMMENTED** [review]
nit comment for improving error message further.

File: python/ray/dag/compiled_dag_node.py (hunk 1 above)
Comment (kevin85421) [reply python/ray/dag/compiled_dag_node.py]: updated https://github.com/ray-project/ray/pull/47757/commits/cc9c8235b92cd32ee15c44ded1eedf158ac61f03

**Comment by kevin85421** [comment]
> Can you create an issue for this? Also can you share me the error message?

@rkooo567 This was the original implementation of this PR, but we decided to raise an exception for leaf nodes after our discussions. Do we still need to open an issue to track it? What error message are you referring to?

File: python/ray/dag/compiled_dag_node.py [hunk 2]
```diff
@@ -946,6 +946,27 @@ def _preprocess(self) -> None:
...
+                len(task.downstream_task_idxs) == 0
+                and not task.dag_node.is_adag_output_node
+            ):
+                leaf_nodes.append(task.dag_node)
+        # Leaf nodes are not allowed because the exception thrown by the leaf
+        # node will not be propagated to the driver.
+        if len(leaf_nodes) != 0:
+            raise ValueError(
+                "Compiled DAG doesn't support leaf nodes that don't have "
+                "downstream nodes and are not output nodes. There are "
```
Comment (ruisearch42) [inline python/ray/dag/compiled_dag_node.py:964]:
```suggestion
                "Compiled DAG doesn't support leaf nodes, i.e., nodes that don't have "
                "downstream nodes and are not output nodes. There are "
```

File: python/ray/dag/tests/experimental/test_accelerated_dag.py [hunk 3]
```diff
@@ -1345,15 +1317,15 @@ def test_compile_twice_with_different_nodes(self, ray_start_regular):
...
+            dag = MultiOutputNode([branch1, branch2])
         compiled_dag = dag.experimental_compile()
         compiled_dag.teardown()
         with pytest.raises(
             ValueError,
             match="The DAG was compiled more than once. The following two "
             "nodes call `experimental_compile`: ",
         ):
-            compiled_dag = branch2.experimental_compile()
+            branch2.experimental_compile()
```
Comment (ruisearch42) [inline python/ray/dag/tests/experimental/test_accelerated_dag.py:1328]: Why is this changed to `branch2`?

File: python/ray/dag/tests/experimental/test_accelerated_dag.py [hunk 4]
```diff
@@ -1888,15 +1860,22 @@ def test_intra_process_channel_with_multi_readers(
 
 
 class TestLeafNode:
+    """
+    Leaf nodes are invalid because the exception thrown by the leaf
+    node will not be propagated to the driver.
```
Comment (ruisearch42) [inline python/ray/dag/tests/experimental/test_accelerated_dag.py:1865]:
```suggestion
    Leaf nodes are not allowed right now because the exception thrown by the leaf
    node will not be propagated to the driver and silently ignored, which is undesired.
```

File: python/ray/dag/tests/experimental/test_accelerated_dag.py [hunk 5]
```diff
@@ -1918,21 +1897,62 @@ def test_leaf_node_two_actors(self, ray_start_regular):
...
+            match=TestLeafNode.LEAF_NODE_EXCEPTION_TEMPLATE.format(num_leaf_nodes=1),
+        ):
+            dag.experimental_compile()
 
-        ref = compiled_dag.execute(10)
-        assert ray.get(ref) == [120, 220]
-        compiled_dag.teardown()
+    def test_multi_leaf_nodes(self, ray_start_regular):
+        """
+        driver -> a.inc -> a.inc (branch 1)
```
Comment (ruisearch42) [inline python/ray/dag/tests/experimental/test_accelerated_dag.py:1915]:
```suggestion
        driver -> a.inc -> a.inc (branch 1, leaf node)
```

File: python/ray/dag/tests/experimental/test_accelerated_dag.py [hunk 6]
```diff
@@ -1918,21 +1897,62 @@ def test_leaf_node_two_actors(self, ray_start_regular):
...
+               |        |
+               |        -> a.inc -> driver
+               |
+               -> a.inc (branch 2)
```
Comment (ruisearch42) [inline python/ray/dag/tests/experimental/test_accelerated_dag.py:1919]:
```suggestion
               -> a.inc (branch 2, leaf node)
```

**Review by ruisearch42 — COMMENTED** [review]
Thinking a bit more on the problem, does it make a real difference if we disallow leaf nodes and force the user to add leaf nodes manually to `MultiOutputNode`?

Using the same example as mentioned, a MultiOutputNode is containing 3 DAG nodes (2 normal DAG nodes + 1 leaf node).

Previously we have:
`x, y = compiled_dag.execute(input_vals)`
and the 3rd return value is silently GC-ed and it calls `get()` underneath.

When the user changes to 
`x, y, z = compiled_dag.execute(input_vals)`

`z` is not used by the user and also discarded, the same GC and `get` happens. Is there a difference here?

File: python/ray/dag/tests/experimental/test_accelerated_dag.py (hunk 3 above)
Comment (kevin85421) [reply python/ray/dag/tests/experimental/test_accelerated_dag.py:1328]: `compiled_dag` is not used below.

**Comment by kevin85421** [comment]
gentle ping - @ruisearch42 @rkooo567 

@rkooo567: I also opened another issue: https://github.com/ray-project/ray/issues/47977.

**Comment by ruisearch42** [comment]
Generally looks good, but after https://github.com/ray-project/ray/pull/47689 , implicitly adding leaf node to MultiOutputNode will work? And do we still need this change?

**Comment by kevin85421** [comment]
> but after https://github.com/ray-project/ray/pull/47689 , implicitly adding leaf node to MultiOutputNode will work? And do we still need this change?

I briefly skimmed through #47689. It will still call get in `__del__` if the graph is not torn down, which will cause `execute` to become a synchronous operation if I understand correctly.

**Review by rkooo567 — APPROVED** [review]
awesome error message!
